services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java

```diff
@@ -110,7 +111,20 @@ public boolean isValid(RealmModel realm, UserModel user, CredentialInput credent
             return true;
         } else if (OTPCredentialModel.TOTP.equals(credentialData.getSubType())) {
             TimeBasedOTP validator = new TimeBasedOTP(credentialData.getAlgorithm(), credentialData.getDigits(), credentialData.getPeriod(), policy.getLookAheadWindow());
-            return validator.validateTOTP(challengeResponse, secretData.getValue().getBytes(StandardCharsets.UTF_8));
+
+            boolean validReusedOtp = policy.isCodeReusable() || !challengeResponse.equals(secretData.getLastValue());
+            if (!validReusedOtp) {
+                logger.debug("It is forbidden to use the same OTP code twice. Please wait for the next one.");
+                return false;
+            }
+
+            if (validator.validateTOTP(challengeResponse, secretData.getValue().getBytes(StandardCharsets.UTF_8))) {
```

Comment:
I think it should be safer to validate codes based on the last successful validation interval rather than the last used code.  Although not a big issue when it happens, it should be possible that the same code is generated for different times. Codes are 6 or 8 digits and I think it should happen.

By doing checks based on the interval, we store the last interval from here https://github.com/keycloak/keycloak/blob/fd2787ae7da15e50a523d52c59617584e6d09cc7/server-spi-private/src/main/java/org/keycloak/models/utils/TimeBasedOTP.java#L83.

Then we change the `validateTOTP` method to accept the last validation interval and introduce a check like:

```
if (currentInterval <= lastValidationInterval) {
    return false;
}
```

Reply:
@pedroigor Thanks a lot for your suggestion! I'm aware that the solution with the last used code is not so elegant and brings other problems to consider.

However, I've been thinking a little bit about it, and the approach, which you've suggested, has some drawbacks as well. The thing is related to the "look around window". The code can be valid even for the following auth windows (the count of valid windows depends on the `lookAroundWindow` field).  I assume that you had in mind, for the successful validation, store either:

1) `lastValidationInterval = currentInterval;` -> Store the last "success" interval and then use the condition, that you've mentioned above. It'd work only without the "look around window" feature. A user can also enter the code from the previous auth window. Moreover, as the code is valid even for the next window, an attacker could use the same code in the next window.

or

2) `lastValidationInterval = currentInterval + lookAroundWindow;` ->  Store the last "success" interval + look around window value (or the evaluation of the valid interval can be done later). It'd mean, that the user is not able to use the code from the next auth window and needs to wait till `currentInterval > lastValidationInterval` - not okay.

--------------------------------------------------------------------
I think the combination of both could work; store the last successful OTP code and the last successful interval. I can assemble a particular solution for that and you can review it later.

If I should elaborate it a little bit more, just feel free to let me know :))

Reply:
Yeah, we need to consider the look ahead window and for that, we should take into account the `adjust` interval when storing the last successful interval.